configfile: "config/config.yaml"

from pathlib import Path
results = Path(config['output_directory'])

include: "rules/create_indices.smk"
include: "rules/align_reads.smk"


# get a list of samples from the samples file
samples = []
with open(config['samples_file'], 'r') as f:
    for line in f:
        sample = line.strip()
        samples.append(sample)


# Rules
# ----------------------------------------------------------------------------------------

rule all:
    input:
        genotyped_cohort = results / 'genotyped_cohort.vcf.gz'

rule copy_config:
    input:
        config = "config/config.yaml"
    output: 
        config = results / "config" / "config.yaml"
    shell:
        "rsync -a {input.config} {output.config}"

rule create_variant_arguments_file:
    input:
        called_haplotypes = expand(results / "aligned_reads" / "{sample}.g.vcf.gz", sample = samples),
    output:
        variant_arguments_file = results / "aligned_reads" / "consolidate_vcf_args.txt"
    run:
        with open(output.variant_arguments_file, 'w') as out:
            for variant in input.called_haplotypes:
                out.write(f"--variant {variant}\n")


rule consolidate gVCFs:
# there is an alternate method using gatk GenomicsDBImport but here we will use gatk CombineGVCFs
# the other method is probably faster for very large datasets
    input:
        variant_arguments_file = results / "aligned_reads" / "consolidate_vcf_args.txt",
        ref_genome = copied_ref_genome,
    output:
        called_merged_haplotypes = results / "aligned_reads" / "called_haplotypes_merged.vcf.gz"
    shell:
        # gvcfs=({input.called_haplotypes})
        '''
         gatk CombineGVCFs \
           --R {input.ref_genome} \
           --arguments_file {input.variant_arguments_file} \
           -O {output.called_merged_haplotypes}
        '''

rule joint_genotyping_general:
    input:
        ref_genome = copied_ref_genome,
        called_merged_haplotypes = results / "aligned_reads" / "called_haplotypes_merged.vcf.gz"
    output:
        genotyped_cohort = results / 'genotyped_cohort.vcf.gz'
    # conda:
    #     "envs/gatk.yaml"
    shell:
        '''
        gatk GenotypeGVCFs \
            -R {input.ref_genome} \
            -V {input.called_merged_haplotypes} \
            -O {output.genotyped_cohort} \
            --force-output-intervals {config[targets_vcf]}
        '''

# rule joint_genotyping_general:
#     input:
#         ref_genome = copied_ref_genome,
#         called_merged_haplotypes = results / "aligned_reads" / "called_haplotypes_merged.vcf.gz"
#     output:
#         genotyped_cohort = results / 'genotyped_cohort.vcf.gz'
#     # conda:
#     #     "envs/gatk.yaml"
#     shell:
#         '''
#         gatk GenotypeGVCFs \
#             -R {input.ref_genome} \
#             -V {input.called_merged_haplotypes} \
#             -O {output.genotyped_cohort}
#         '''

# rule joint_genotyping_targeted:
#     input:
#         ref_genome = copied_ref_genome,
#         called_merged_haplotypes = results / "called_haplotypes" / "called_haplotypes_merged.vcf.gz"
#     output:
#         genotyped_cohort_targeted = results / 'genotyped_cohort_targeted.vcf.gz'
#     # conda:
#     #     "envs/gatk.yaml"
#     shell:
#         '''
#         gatk GenotypeGVCFs \
#             -R {input.ref_genome} \
#             -V {input.called_merged_haplotypes} \
#             -O {output.genotyped_cohort_targeted} \
#             --alleles config/targets_test.vcf \
#             --genotype-germline-sites true
#         '''        
#             # --include-non-variant-sites
#             # -L config/targets.vcf \

# rule merge_after_joint_genotyping:
#     input:
#         genotyped_cohort = results / 'genotyped_cohort.vcf.gz',
#         genotyped_cohort_targeted = results / 'genotyped_cohort_targeted.vcf.gz'
#     output:
#         genotyped_cohort_merged = results / 'genotyped_cohort_merged.vcf'
#     shell:
#         "bcftools concat {input.genotyped_cohort} {input.genotyped_cohort_targeted} "
#             "-o {output.genotyped_cohort_merged} "
#             "-aD" #allow for overlapping sites and remove any duplicates
