# import config
# -----------------------------------------------------------------
configfile: "config/config.yaml"

# define variables
# ---------------------------------------------------------------
from pathlib import Path
import subprocess

samples = [
    x.name.removesuffix(config['R1_suffix']) for x in
    Path(config['sample_reads_folder']).glob("*"+config['R1_suffix'])
]

results = Path(config['output_directory'])

if config['amplicon_sequencing']:
    bqsr_input_path =  results / "temp" / "aligned_reads"
elif not config['amplicon_sequencing']:
    bqsr_input_path =  results / "temp" / "marked_duplicates"

# include external modules
# -----------------------------------------------------------------------------------
module snp_eff_annotation:
    snakefile:
        github("simkin-bioinformatics/snp_eff_annotation", path = "/workflow/Snakefile", commit = "329cedaf5a2ca47b0e62ac331e3d1813b57b8d21")
        # "/home/charlie/snp_eff_annotation/workflow/Snakefile"
    config: config

use rule * from snp_eff_annotation exclude copy_config as snpeff_*

use rule run_snp_eff from snp_eff_annotation as snpeff_run_snp_eff with:
    input:
        snp_eff_config = results / "snpEff_processing" / "snpEff.config",
        snp_eff_database = results / "snpEff_processing"/ "data" / config['genome_database_name'] / "sequence.bin",
        merged_vcf = results / 'genotyped_cohort.vcf.gz',

# Rules
# ----------------------------------------------------------------------------------------
include: "rules/create_indices.smk"
include: "rules/call_haplotypes.smk"
include: "rules/joint_genotyping.smk"

onsuccess:
    subprocess.run(["rm", "-rf", results / "temp"])
    subprocess.run(["rm", "-rf", results / "snpEff_processing"])

rule all:
    input:
        rules.snpeff_all.input,
        copied_config = results / "config" / "config.yaml",
        copied_targets_tsv = results / "config" / Path(config['targets_tsv']).name,
    default_target: True

rule copy_config:
    input:
        config = "config/config.yaml",
        targets_tsv = config['targets_tsv']
    output:
        config = results / "config" / "config.yaml",
        targets_tsv = results / "config" / Path(config['targets_tsv']).name,
    shell:
        "rsync -a {input.config} {output.config} && "
        "rsync -a {input.targets_tsv} {output.targets_tsv}"
