configfile: "config/config.yaml"

from pathlib import Path

samples = [
    x.name.removesuffix(config['R1_suffix']) for x in
    Path(config['sample_reads_folder']).glob("*"+config['R1_suffix'])
]

results = Path(config['output_directory'])

include: "rules/create_indices.smk"
include: "rules/align_reads.smk"

snp_eff_copied = results / Path(config['snp_eff_folder']).name
module snp_eff_annotation:
    snakefile:
        "/home/charlie/snp_eff_annotation/workflow/Snakefile"
    config: config

use rule * from snp_eff_annotation exclude copy_config as snpeff_*



# get a list of samples from the samples file
# samples = []
# with open(config['samples_file'], 'r') as f:
#     for line in f:
#         sample = line.strip()
#         samples.append(sample)


# Rules
# ----------------------------------------------------------------------------------------

rule all:
    input:
        # genotyped_cohort = results / 'genotyped_cohort.vcf.gz'
        coverage_AA=results / "AA_tables" / "coverage_AA_table.csv",
        reference_AA=results / "AA_tables" / "reference_AA_table.csv",
        alternate_AA=results / "AA_tables" / "alternate_AA_table.csv",
        config = results / "config" / "config.yaml",
        targets_vcf = results / "config" / Path(config['targets_vcf']).name,
        targets_tsv = results / "config" / Path(config['targets_tsv']).name,
    default_target: True

rule copy_config:
    input:
        config = "config/config.yaml",
        targets_vcf = config['targets_vcf'],
        targets_tsv = config['targets_tsv']
    output:
        config = results / "config" / "config.yaml",
        targets_vcf = results / "config" / Path(config['targets_vcf']).name,
        targets_tsv = results / "config" / Path(config['targets_tsv']).name,
    shell:
        "rsync -a {input.config} {output.config} && "
        "rsync -a {input.targets_vcf} {output.targets_vcf} && "
        "rsync -a {input.targets_tsv} {output.targets_tsv}"

rule create_variant_arguments_file:
    input:
        called_haplotypes = expand(results / "aligned_reads" / "{sample}.g.vcf.gz", sample = samples),
    output:
        variant_arguments_file = results / "aligned_reads" / "consolidate_vcf_args.txt"
    run:
        with open(output.variant_arguments_file, 'w') as out:
            for variant in input.called_haplotypes:
                out.write(f"--variant {variant}\n")


rule consolidate gVCFs:
# there is an alternate method using gatk GenomicsDBImport but here we will use gatk CombineGVCFs
# the other method is probably faster for very large datasets
    input:
        variant_arguments_file = results / "aligned_reads" / "consolidate_vcf_args.txt",
        ref_genome = copied_ref_genome,
    output:
        called_merged_haplotypes = results / "aligned_reads" / "called_haplotypes_merged.vcf.gz"
    shell:
        # gvcfs=({input.called_haplotypes})
        '''
         gatk CombineGVCFs \
           --R {input.ref_genome} \
           --arguments_file {input.variant_arguments_file} \
           -O {output.called_merged_haplotypes}
        '''

rule joint_genotyping_general:
    input:
        ref_genome = copied_ref_genome,
        called_merged_haplotypes = results / "aligned_reads" / "called_haplotypes_merged.vcf.gz"
    output:
        genotyped_cohort = results / 'genotyped_cohort.vcf.gz'
    shell:
        '''
        gatk GenotypeGVCFs \
            -R {input.ref_genome} \
            -V {input.called_merged_haplotypes} \
            -O {output.genotyped_cohort} \
            --force-output-intervals {config[targets_vcf]}
        '''
